src/ws/server.rs

```diff
@@ -21,476 +21,134 @@
 // SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 // CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 // OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
-// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// IN background_task WITH THE SOFTWARE OR THE USE OR OTHER
 // DEALINGS IN THE SOFTWARE.
 
-use crate::types::error::Error;
-use crate::types::jsonrpc::{self, JsonValue};
-use crate::ws::raw::{RawServer, RawServerEvent, RawServerRequestId, RawServerSubscriptionId};
-use crate::ws::transport::WsTransportServer;
+use std::sync::Arc;
+use serde::{Serialize, Deserialize};
+use serde_json::value::RawValue;
+use futures::io::{BufReader, BufWriter};
+use soketto::handshake::{Server as SokettoServer, server::Response};
+use tokio::stream::StreamExt;
+use tokio::sync::mpsc;
+use tokio_util::compat::Tokio02AsyncReadCompatExt;
+use thiserror::Error;
+use rustc_hash::FxHashMap;
 
-use futures::{channel::mpsc, future::Either, pin_mut, prelude::*};
-use parking_lot::Mutex;
-use std::{
-	collections::{HashMap, HashSet},
-	convert::TryFrom,
-	error,
-	net::SocketAddr,
-	sync::{atomic, Arc},
-};
+use crate::ws::types::{JsonRpcRequest, JsonRpcResponse, TwoPointZero};
 
-/// Server that can be cloned.
-///
-/// > **Note**: This struct is designed to be easy to use, but it works by maintaining a background
-/// >           task running in parallel. If this is not desirable, you are encouraged to use the
-/// >           [`RawServer`] struct instead.
-#[derive(Clone)]
-pub struct Server {
-	/// Channel to send requests to the background task.
-	to_back: mpsc::UnboundedSender<FrontToBack>,
-	/// List of methods (for RPC queries, subscriptions, and unsubscriptions) that have been
-	/// registered. Serves no purpose except to check for duplicates.
-	registered_methods: Arc<Mutex<HashSet<String>>>,
-	/// Next unique ID used when registering a subscription.
-	next_subscription_unique_id: Arc<atomic::AtomicUsize>,
-	/// Local socket address of the transport server.
-	local_addr: SocketAddr,
-}
-
-/// Notification method that's been registered.
-pub struct RegisteredNotification {
-	/// Receives notifications that the client sent to us.
-	queries_rx: mpsc::Receiver<jsonrpc::Params>,
-}
+type Methods = FxHashMap<&'static str, Box<dyn Send + Sync + Fn(Option<&RawValue>, &str, &mpsc::UnboundedSender<String>) -> anyhow::Result<()>>>;
```

Comment:
`Box<dyn Fn()>` here is syntactic sugar `Box<dyn Fn()  + 'static>` so we need to add explicit lifetime here for this to work.

Did you try that @dvdplm?